### PR TITLE
Fix indentation for function argument in call expression

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes a vsfeedback issue:
https://developercommunity.visualstudio.com/content/problem/40788/javascript-when-using-the-universal-module-definit.html

Concretely, when using writing UMD-modules, the user wraps a factory method in an iife. the particular placement of newlines when writing the pattern would trigger an edge-case in our indentation API.

In order to determine indentation, we walk up the AST from the starting position and compare each node with its parent in turn. Roughly-speaking, if a node and its parent don't start on the same line, we would increase the indentation we compute accordingly.

### Problem
In our AST representation, the parent of a call argument is the call-expression, not the argument list. But we would like to increase indentation based on the relationship between an argument and its argument-list, so we spoof the starting position of the (parent) call-expression (see [`getParentStart`](https://github.com/Microsoft/TypeScript/blob/master/src/services/formatting/smartIndenter.ts#L186)). But then the spoofed start-value could cause a problem when we compare the start position of the call-expression to *its* parent (in the iife case, an expression statement). This adds an extra level of indentation.

### Solution
When we are in an argument, unspoof the starting position of the enclosing call expression after applying indentation for the argument.

#

Note, with this change we conservatively only unspoof the start position for call expressions. I haven't thoroughly considered the implications of resetting the start position for other expressions enclosing lists of elements (eg: type parameters in function declarations, elements in named imports or exports, etc.), and users haven't (to my knowledge) complained, so we leave the behavior unchanged for these cases.

### CODE REVIEW

File: src/services/formatting/smartIndenter.ts [hunk 1]
```diff
@@ -268,6 +275,16 @@ namespace ts.formatting {
             return sourceFile.getLineAndCharacterOfPosition(n.getStart(sourceFile));
         }
 
+        export function isArgumentAndStartLineOverlapsExpressionBeingCalled(parent: Node, child: Node, childStartLine: number, sourceFile: SourceFileLike): boolean {
+            if (!(isCallExpression(parent) && contains(parent.arguments, child))) {
```
Comment (amcasey) [inline src/services/formatting/smartIndenter.ts]: Would a similar issue arise for an index expression in an array access?

File: src/services/utilities.ts [hunk 2]
```diff
@@ -402,8 +402,11 @@ namespace ts {
         return start < end;
     }
 
+    /**
+     * Assumes `candidate.start <= position` holds.
```
Comment (aozgaa) [inline src/services/utilities.ts]: We don't have any conditional compilation. That is, if we wrote `Debug.assert(...)`, the enclosed expression would be run every time.

Since the assumed value isn't checked for performance reasons, it would defeat the purpose to add an assert that always runs.

File: src/services/formatting/smartIndenter.ts [hunk 3]
```diff
@@ -9,18 +9,21 @@ namespace ts.formatting {
...
 
         /**
-         * Computed indentation for a given position in source file
-         * @param position - position in file
-         * @param sourceFile - target source file
-         * @param options - set of editor options that control indentation
-         * @param assumeNewLineBeforeCloseBrace - false when getIndentation is called on the text from the real source file.
-         * true - when we need to assume that position is on the newline. This is usefult for codefixes, i.e.
+         * @param assumeNewLineBeforeCloseBrace
+         * `false` when called on text from a real source file.
```
Comment (aozgaa) [inline src/services/formatting/smartIndenter.ts]: Being called with a synthesized AST. The code-fixes in particular do this, and invoke the API's in `textChanges.ts`.

File: tests/cases/fourslash/indentationInAmdIife.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+////     function () {/*8_1*/
+////     });
+//// (foo)(1, function() 
+//// {/*4_2*/
+//// });
+//// 
+//// // Contains line-breaks in the expression part of the call expression.
+////
+//// (
+////     foo)(1, function () {/*4_1*/
```
Comment (aozgaa) [inline tests/cases/fourslash/indentationInAmdIife.ts]: This is a consequence of the algorithm. We can fix this later if it crops up, but it's tricky to change without special-casing because we are simulating the argument-list as a parent of the lambda.

File: src/services/formatting/smartIndenter.ts [hunk 5]
```diff
@@ -174,22 +179,24 @@ namespace ts.formatting {
...
+                // Update `current` and `parent`.
+
+                const useTrueStart =
+                    isArgumentAndStartLineOverlapsExpressionBeingCalled(parent, current, currentStart.line, sourceFile);
+
                 current = parent;
-                currentStart = parentStart;
                 parent = current.parent;
+                currentStart = useTrueStart ? sourceFile.getLineAndCharacterOfPosition(current.getStart()) : containingListOrParentStart;
+                containingListOrParentStart = undefined;
```
Comment (billti) [inline src/services/formatting/smartIndenter.ts]: What's the point or setting this to undefined at the end? Looks like it's always reassigned inside the loop, and unused outside of it.

File: src/services/formatting/smartIndenter.ts [hunk 6]
```diff
@@ -9,18 +9,21 @@ namespace ts.formatting {
...
-         * when inserting some text after open brace we would like to get the value of indentation as if newline was already there.
-         * However by default indentation at position | will be 0 so 'assumeNewLineBeforeCloseBrace' allows to override this behavior,
+         * ```
+         * with `position` at `|`.
+         *
+         * When inserting some text after an open brace, we would like to get indentation as if a newline was already there.
+         * By default indentation at `position` will be 0 so 'assumeNewLineBeforeCloseBrace' overrides this behavior.
          */
-        export function getIndentation(position: number, sourceFile: SourceFile, options: EditorSettings, assumeNewLineBeforeCloseBrace = false): number {
+        export function getIndentationAtPosition(position: number, sourceFile: SourceFile, options: EditorSettings, assumeNewLineBeforeCloseBrace = false): number {
```
Comment (billti) [inline src/services/formatting/smartIndenter.ts]: I know the namespace is marked 'internal', but this is still an exported function, and someone may be using it. I'd avoid this rename unless there's a clear need for it.

File: src/services/formatting/smartIndenter.ts [hunk 7]
```diff
@@ -174,22 +179,24 @@ namespace ts.formatting {
                     indentationDelta += options.indentSize;
                 }
 
+                // Update `current` and `parent`.
```
Comment (billti) [inline src/services/formatting/smartIndenter.ts]: nit: This is kind of a 'duh' comment. You can see they are both being assigned to in the following lines.

File: src/services/formatting/smartIndenter.ts [hunk 8]
```diff
@@ -174,22 +179,24 @@ namespace ts.formatting {
                     indentationDelta += options.indentSize;
                 }
 
+                // Update `current` and `parent`.
+
+                const useTrueStart =
```
Comment (billti) [inline src/services/formatting/smartIndenter.ts]: You should put some of the explanation you have written-up on this pull request into the code as comments. The reasoning behind some of this code is really hard to follow without understanding the problems it is trying to solve.

File: src/services/utilities.ts [hunk 9]
```diff
@@ -407,8 +407,11 @@ namespace ts {
         return start < end;
     }
 
+    /**
+     * Assumes `candidate.start <= position` holds.
+     */
     export function positionBelongsToNode(candidate: Node, position: number, sourceFile: SourceFile): boolean {
-        return candidate.end > position || !isCompletedNode(candidate, sourceFile);
+        return position < candidate.end || !isCompletedNode(candidate, sourceFile);
```
Comment (billti) [inline src/services/utilities.ts]: I can tell you have a preference for ordering comparisons as I've seen you do this in several pull requests (probably based on your math background :wink:), but please don't go changing existing expressions unless you are re-writing that code anyway, they are unreadable as written, or they are incorrect.

File: tests/cases/fourslash/indentationInAmdIife.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+////     (1, function () {/*8_2*/
+////     });
+//// (foo
+//// )(1,
+////     function () {/*8_3*/
+////     });
+//// (foo
+//// )(1, function()
+//// {/*4_4*/
+//// });
```
Comment (billti) [inline tests/cases/fourslash/indentationInAmdIife.ts]: Do we have existing coverage for other areas that could regress (e.g. typing newlines inside other expressions, arrays, objects, argument lists, etc.)? We should make sure these don't regress. (i.e. if there is no existing coverage, we should add some basic tests with this change).

Comment (aozgaa) [reply tests/cases/fourslash/indentationInAmdIife.ts]: tests added.

File: src/services/formatting/smartIndenter.ts (hunk 8 above)
Comment (aozgaa) [reply src/services/formatting/smartIndenter.ts]: added.

File: src/services/formatting/smartIndenter.ts (hunk 7 above)
Comment (aozgaa) [reply src/services/formatting/smartIndenter.ts]: removed

File: src/services/formatting/smartIndenter.ts (hunk 6 above)
Comment (aozgaa) [reply src/services/formatting/smartIndenter.ts]: reverted

File: src/services/formatting/smartIndenter.ts (hunk 5 above)
Comment (aozgaa) [reply src/services/formatting/smartIndenter.ts]: It was placed to emphasize that the value of `containingListOrParentStart` is meaningless at the start of the loop, but this could be seen by just looking at what happens in the first iteration.

Comment (aozgaa) [reply src/services/formatting/smartIndenter.ts]: Removed.

File: src/services/formatting/smartIndenter.ts (hunk 1 above)
Comment (aozgaa) [reply src/services/formatting/smartIndenter.ts]: We could keep looking into awkward situations ad infinitum, but we would increasingly approach patterns that don't resemble code user write. To avoid a rabbit hole, I'd prefer to keep this change conservative and get it in. If we run into issues in the course of use or get customer feedback, we can adjust our indentation then.

This change was designed to minimally effect indentation except for the args-in-iife case described in the issue. The supplementary tests add (which pass unchanged from master) provide some confirmation this actually holds.
